components/web_discovery/browser/credential_manager.h

```diff
@@ -0,0 +1,132 @@
+// b) "signing": uses the previously acquired credentials to sign submissions
+//    which is required in order for the servers to accept the request.
+class CredentialManager : public CredentialSigner {
+ public:
+  CredentialManager(PrefService* profile_prefs,
+                    network::SharedURLLoaderFactory* shared_url_loader_factory,
+                    const ServerConfigLoader* server_config_loader);
+  ~CredentialManager() override;
+
+  CredentialManager(const CredentialManager&) = delete;
+  CredentialManager& operator=(const CredentialManager&) = delete;
+
+  // Acquires credentials for all dates/"group public keys" published in
+  // the server config, if not stored already.
+  void JoinGroups();
+
+  // CredentialSigner:
+  bool CredentialExistsForToday() override;
+
+  bool Sign(std::vector<const uint8_t> msg,
+            std::vector<const uint8_t> basename,
+            SignCallback callback) override;
+
+  // Uses a fixed seed in the anonymous credential manager
+  // to provide deterministic credentials & signatures which
+  // are useful for testing.
+  void UseFixedSeedForTesting();
+
+ private:
+  bool LoadRSAKey();
+  bool GenerateRSAKey();
+  void OnNewRSAKey(std::unique_ptr<RSAKeyInfo> key_info);
+
+  void StartJoinGroup(const std::string& date,
+                      const std::string& group_pub_key_b64);
+
+  void OnJoinRequestReady(
+      std::string date,
+      std::vector<const uint8_t> group_pub_key,
+      std::optional<GenerateJoinRequestResult> generate_join_result);
+
+  void OnJoinResponse(std::string date,
+                      std::vector<const uint8_t> group_pub_key,
+                      std::vector<const uint8_t> gsk,
+                      std::optional<std::string> response_body);
+  void HandleJoinResponseStatus(const std::string& date, bool result);
+  bool ProcessJoinResponse(const std::string& date,
+                           const std::vector<const uint8_t>& group_pub_key,
+                           const std::vector<const uint8_t>& gsk,
+                           const std::optional<std::string>& response_body);
+  void OnCredentialsReady(std::string date,
+                          std::vector<const uint8_t> gsk,
+                          std::optional<std::string> credentials);
+
+  void OnSignResult(std::string credential_date,
+                    SignCallback callback,
+                    std::optional<std::vector<const uint8_t>> signed_message);
+
+  raw_ptr<PrefService> profile_prefs_;
+  raw_ptr<network::SharedURLLoaderFactory> shared_url_loader_factory_;
+  raw_ptr<const ServerConfigLoader> server_config_loader_;
+
+  GURL join_url_;
+  base::flat_map<std::string, std::unique_ptr<network::SimpleURLLoader>>
+      join_url_loaders_;
+  net::BackoffEntry backoff_entry_;
+  base::WallClockTimer retry_timer_;
+
+  scoped_refptr<base::SequencedTaskRunner> sequenced_task_runner_;
+
+  std::unique_ptr<rust::Box<anonymous_credentials::CredentialManager>,
```

Comment:
why are you wrapping a rust::Box inside a unique_ptr?

Reply:
updated to use `base::SequenceBound`